drivers/dac/dac_gd32.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2021 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_dac
+
+#include <errno.h>
+#include <drivers/pinctrl.h>
+#include <drivers/dac.h>
+#include <soc.h>
+
+#define LOG_LEVEL CONFIG_DAC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dac_gd32);
+
+/**
+ * Most of GD32 series SOCs' DAC have 2 output, DAC0 and DAC1.
+ * But some SOC, like GD32F350 only have one output, DAC0.
+ */
+#if !defined(CONFIG_SOC_GD32F350)
+#define DAC_GD32_OUT_NUM  2
+#define DAC_GD32_OUT_DUAL 1
+#else
+#define DAC_GD32_OUT_NUM  1
+#endif /* !CONFIG_SOC_GD32F350 */
+
+#define DAC_GD32_DEFAULT_VALUE 0
+
+struct dac_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct dac_gd32_data {
+	uint8_t resolutions[DAC_GD32_OUT_NUM];
+};
+
+static void dac_gd32_write(struct dac_gd32_data *data,
+		    uint8_t dacx, uint32_t value)
+{
+	uint32_t align;
+
+	if (data->resolutions[dacx] == 8) {
+		align = DAC_ALIGN_8B_R;
+	} else {
+		align = DAC_ALIGN_12B_R;
+	}
+
+#ifdef DAC_GD32_OUT_DUAL
+	dac_data_set(dacx, align, value);
+#else
+	dac_data_set(align, value);
+#endif /* DAC_GD32_OUT_DUAL */
+}
+
+static int dac_gd32_channel_setup(const struct device *dev,
+			   const struct dac_channel_cfg *channel_cfg)
+{
+	struct dac_gd32_data *data = dev->data;
+	uint8_t dacx = channel_cfg->channel_id;
+
+	if (dacx >= DAC_GD32_OUT_NUM) {
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->resolution == 8 ||
+	    channel_cfg->resolution == 12) {
+		data->resolutions[dacx] = channel_cfg->resolution;
+	} else {
+		return -ENOTSUP;
+	}
+
+#if DAC_GD32_OUT_DUAL
+	dac_disable(dacx);
+	dac_gd32_write(data, dacx, DAC_GD32_DEFAULT_VALUE);
+	dac_enable(dacx);
+#else
+	dac_deinit();
+	dac_gd32_write(data, dacx, DAC_GD32_DEFAULT_VALUE);
+	dac_enable();
+#endif /* DAC_GD32_OUT_DUAL */
```

Comment:
consider using HAL register-based macros if HAL API is not consistent with namings (`dac_disable` vs `dac_deinit`).

Reply:
done.